File: Java/kadane_algorithm.java
import java.util.*;
public class Main {
    public static void main(String args[]) {
        Scanner sc= new Scanner(System.in);
        int size=sc.nextInt();
        int[] arr=new int[size];
         for(int i=0;i<arr.length;i++){
             arr[i]=sc.nextInt();
         }
        int ans=maximumSumSubarray(arr);
        System.out.println("Maximum Sum is "+ans);
   }
   public static int maximumSumSubarray(int[] arr){
       int max=Integer.MIN_VALUE;
         int temp=0;
       for(int i=0;i<arr.length;i++){
         temp+=arr[i];
         if(max < temp){
             max=temp;
         }
         if(temp < 0){
             temp=0;
         }
       }
       return max;
   }
}

File: Python/kadane_algorithm.py
# Kadane's Algorithm gives the maximum sum of a subarray in the given array.
# This function takes a python list as argument and return a number which is the
# max sum of any subarray of this list
def kadanes_algorithm(arr):
    maxTillNow = float('-inf')
    maxHere = 0

    for i in range(len(arr)):
        maxHere += arr[i]
        if maxTillNow < maxHere:
            maxTillNow = maxHere
        if maxHere < 0:
            maxHere = 0

    return maxTillNow


print("Enter space seperated list of numbers: ")
arr = list(map(int, input().split()))
print(kadanes_algorithm(arr))
"""
    Time Complexity: O(n)
    Space Complexity: O(n)

    INPUT:
    1 4 -5 6 -1 -2 7 0
    OUTPUT:
    10
"""
